File: kirun-js/__tests__/engine/function/system/array/IndexOfTest.ts
import { IndexOf } from '../../../../../src/engine/function/system/array/IndexOf';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

test('Index of Test 1', () => {
    let ind: IndexOf = new IndexOf();

    let array: string[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let find: string = 'with';

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters().setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
            [IndexOf.PARAMETER_INT_FIND_FROM.getParameterName(), 2],
        ]),
    );
    expect(
        ind.execute(fep).allResults()[0].getResult().get(IndexOf.EVENT_RESULT_INTEGER.getName()),
    ).toBe(12);
});

test('Index of Test 2', () => {
    let ind: IndexOf = new IndexOf();

    let array: string[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let find: string = 'with';

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters().setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
            [IndexOf.PARAMETER_INT_FIND_FROM.getParameterName(), -2],
        ]),
    );

    expect(() => ind.execute(fep)).toThrow();

    fep.setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), []],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
            [IndexOf.PARAMETER_INT_FIND_FROM.getParameterName(), -2],
        ]),
    );
    expect(ind.execute(fep).allResults()[0].getResult().get(IndexOf.EVENT_INDEX.getName())).toBe(
        -1,
    );
});

test('Index of Test 3', () => {
    let ind: IndexOf = new IndexOf();

    let array: string[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let find: string = 'witah';

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters().setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
            [IndexOf.PARAMETER_INT_FIND_FROM.getParameterName(), 0],
        ]),
    );

    expect(
        ind.execute(fep).allResults()[0].getResult().get(IndexOf.EVENT_RESULT_INTEGER.getName()),
    ).toBe(-1);
});

test('Index of Test 4', () => {
    let ind: IndexOf = new IndexOf();

    let array: string[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let array2: string[] = [];

    array2.push('test');
    array2.push('Driven');
    array2.push('developement');
    array2.push('I');
    array2.push('am');
    array2.push('using');
    array2.push('eclipse');
    array2.push('I');
    array2.push('to');
    array2.push('test');
    array2.push('the');
    array2.push('changes');
    array2.push('with');

    let array3: string[] = [];
    array3.push('test');
    array3.push('Driven');
    array3.push('developement');
    array3.push('I');
    array3.push('am');
    array3.push('using');
    array3.push('eclipse');
    array3.push('I');
    array3.push('to');
    array3.push('test');
    array3.push('the');
    array3.push('changes');
    array3.push('with');
    array3.push('test');
    array3.push('Driven');
    array3.push('developement');

    let array4: string[] = [];
    array4.push('test');
    array4.push('Driven');
    array4.push('developement');
    array4.push('I');
    array4.push('am');
    array4.push('using');
    array4.push('eclipse');
    array4.push('I');
    array4.push('to');

    let arr: any[] = [];
    arr.push(array3);
    arr.push(array);
    arr.push(array3);
    arr.push(array2);
    arr.push(array4);
    arr.push(array);

    let find: any = array;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters().setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
            [IndexOf.PARAMETER_INT_FIND_FROM.getParameterName(), 2],
        ]),
    );

    expect(
        ind.execute(fep).allResults()[0].getResult().get(IndexOf.EVENT_RESULT_INTEGER.getName()),
    ).toBe(2);
});

test('indexof test 5', () => {
    let ind: IndexOf = new IndexOf();

    let array1: string[] = [];

    array1.push('test');
    array1.push('Driven');
    array1.push('developement');
    array1.push('I');
    array1.push('am');

    let js1: object = {
        boolean: false,
        array: array1,
        char: 'o',
    };

    let js2: object = {
        boolean: false,
        array: array1,
        char: 'asd',
    };

    let js3: object = {
        array: array1,
    };

    let js4: object = {
        boolean: false,
        array: array1,
        char: 'o',
    };

    let arr: any[] = [js1, js2, js1, js3, js3, js4, js1];
    let find: any = js4;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters().setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), find],
        ]),
    );

    expect(
        ind.execute(fep).allResults()[0].getResult().get(IndexOf.EVENT_RESULT_INTEGER.getName()),
    ).toBe(5);

    fep.setArguments(
        new Map<string, any>([
            [IndexOf.PARAMETER_ARRAY_SOURCE.getParameterName(), arr],
            [IndexOf.PARAMETER_ANY_NOT_NULL.getParameterName(), null],
        ]),
    );
    expect(() => ind.execute(fep)).toThrow();
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/IndexOfTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class IndexOfTest {

	@Test
	void test() {
		var array = new JsonArray();

		array.add("test");
		array.add("Driven");
		array.add("developement");
		array.add("I");
		array.add("am");
		array.add("using");
		array.add("eclipse");
		array.add("I");
		array.add("to");
		array.add("test");
		array.add("the");
		array.add("changes");
		array.add("with");
		array.add("test");
		array.add("Driven");
		array.add("developement");

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", array, "elementObject", new JsonPrimitive("with"), "findFrom",
						new JsonPrimitive(2)));

		IndexOf ind = new IndexOf();

		StepVerifier.create(ind.execute(fep))
				.expectNextMatches(result -> {
					return result.next()
							.getResult()
							.get("result")
							.equals(new JsonPrimitive(12));
				})
				.verifyComplete();

	}

	@Test
	void test2() {
		var array = new JsonArray();

		array.add("test");
		array.add("Driven");
		array.add("developement");
		array.add("I");
		array.add("am");
		array.add("using");
		array.add("eclipse");
		array.add("I");
		array.add("to");
		array.add("test");
		array.add("the");
		array.add("changes");
		array.add("with");
		array.add("test");
		array.add("Driven");
		array.add("developement");

		IndexOf ind = new IndexOf();

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", array, "elementObject", new JsonPrimitive("developement"), "findFrom",
						new JsonPrimitive(-2)));

		StepVerifier.create(ind.execute(fep))
				.expectError(KIRuntimeException.class)
				.verify();

		ReactiveFunctionExecutionParameters fep1 = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", new JsonArray(), "elementObject", new JsonPrimitive("developement"),
						"findFrom", new JsonPrimitive(2)));

		StepVerifier.create(ind.execute(fep1))
				.expectNextMatches(result -> {
					return result.next()
							.getResult()
							.get("result")
							.equals(new JsonPrimitive(-1));
				})
				.verifyComplete();
	}

	@Test
	void test3() {
		var array = new JsonArray();

		array.add("test");
		array.add("Driven");
		array.add("developement");
		array.add("I");
		array.add("am");
		array.add("using");
		array.add("eclipse");
		array.add("I");
		array.add("to");
		array.add("test");
		array.add("the");
		array.add("changes");
		array.add("with");
		array.add("test");
		array.add("Driven");
		array.add("developement");

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", array, "elementObject", new JsonPrimitive("developedment"), "findFrom",
						new JsonPrimitive(2)));

		IndexOf ind = new IndexOf();

		StepVerifier.create(ind.execute(fep))
				.expectNextMatches(result -> {
					return result.next()
							.getResult()
							.get("result")
							.equals(new JsonPrimitive(-1));
				})
				.verifyComplete();
	}

	@Test
	void test4() {

		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");
		array1.add("using");
		array1.add("eclipse");
		array1.add("I");
		array1.add("to");
		array1.add("test");
		array1.add("the");
		array1.add("changes");
		array1.add("with");
		array1.add("test");
		array1.add("Driven");
		array1.add("developement");

		var array2 = new JsonArray();

		array2.add("test");
		array2.add("Driven");
		array2.add("developement");
		array2.add("I");
		array2.add("am");
		array2.add("using");
		array2.add("eclipse");
		array2.add("I");
		array2.add("to");
		array2.add("test");
		array2.add("the");
		array2.add("changes");
		array2.add("with");

		var array3 = new JsonArray();

		array3.add("test");
		array3.add("Driven");
		array3.add("developement");
		array3.add("I");
		array3.add("am");
		array3.add("using");
		array3.add("eclipse");
		array3.add("I");
		array3.add("to");
		array3.add("test");
		array3.add("the");
		array3.add("changes");
		array3.add("with");
		array3.add("test");
		array3.add("Driven");
		array3.add("developement");

		var array4 = new JsonArray();

		array4.add("test");
		array4.add("Driven");
		array4.add("developement");
		array4.add("I");
		array4.add("am");
		array4.add("using");
		array4.add("eclipse");
		array4.add("I");
		array4.add("to");

		var arr = new JsonArray();
		arr.add(array1);
		arr.add(array3);
		arr.add(array2);
		arr.add(array4);
		arr.add(array1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "elementObject", array1, "findFrom", new JsonPrimitive(0), "length",
						new JsonPrimitive(2)));

		IndexOf ind = new IndexOf();

		StepVerifier.create(ind.execute(fep))
				.expectNextMatches(result -> {
					return result.next()
							.getResult()
							.get("result")
							.equals(new JsonPrimitive(0));
				})
				.verifyComplete();

	}

	@Test
	void test5() {

		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");

		var js1 = new JsonObject();

		js1.addProperty("boolean", false);
		js1.add("array", array1);
		js1.addProperty("char", 'o');

		var js2 = new JsonObject();

		js2.addProperty("boolean", false);
		js2.add("array", array1);
		js2.addProperty("char", "asd");

		var js3 = new JsonObject();
		js3.add("array", array1);

		var js4 = new JsonObject();

		js4.addProperty("boolean", false);
		js4.add("array", array1);
		js4.addProperty("char", 'o');

		var arr = new JsonArray();

		arr.add(js1);
		arr.add(js2);
		arr.add(js3);
		arr.add(js4);
		arr.add(js1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "elementObject", js4, "findFrom", new JsonPrimitive(0)));

		IndexOf ind = new IndexOf();

		StepVerifier.create(ind.execute(fep))
				.expectNextMatches(result -> {
					return result.next()
							.getResult()
							.get("result")
							.equals(new JsonPrimitive(3));
				})
				.verifyComplete();
	}

	@Test
	void test6() {
		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");

		var js1 = new JsonObject();

		js1.addProperty("boolean", false);
		js1.add("array", array1);
		js1.addProperty("char", 'o');

		var js2 = new JsonObject();

		js2.addProperty("boolean", false);
		js2.add("array", array1);
		js2.addProperty("char", "asd");

		var js3 = new JsonObject();
		js3.add("array", array1);

		var js4 = new JsonObject();

		js4.addProperty("boolean", false);
		js4.add("array", array1);
		js4.addProperty("char", 'o');

		var arr = new JsonArray();

		arr.add(js1);
		arr.add(js2);
		arr.add(js3);
		arr.add(js4);
		arr.add(js1);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of("source", arr, "elementObject", arr, "findFrom", JsonNull.INSTANCE));

		IndexOf ind = new IndexOf();

		StepVerifier.create(ind.execute(fep)
				.map(FunctionOutput::next)
				.map(EventResult::getResult)
				.map(e -> e.get("result")))
				.expectNext(new JsonPrimitive(-1))
				.verifyComplete();
	}

}
